tests/unit/keyinput/test_basekeyparser.py

```diff
@@ -90,6 +99,49 @@ def test_read_config_valid(self):
         assert 'foo' in kp.bindings
         assert 'ctrl+x' in kp.special_bindings
 
+    def test_on_keyconfig_changed_mode_none(self):
+        """Test the changes in config with _modename = None."""
+        kp = basekeyparser.BaseKeyParser(0, supports_count=False,
+                                         supports_chains=False)
+        assert kp._modename is None
+
+        # No config set so self._modename is None
+        with pytest.raises(AssertionError) as excinfo:
+            kp.on_keyconfig_changed('normal')
+        expected_text = "on_keyconfig_changed called but no section defined!"
+        assert str(excinfo.value) == expected_text
+
+    def test_on_keyconfig_changed_mode_normal(self):
+        """Test the changes in config with _modename set."""
+        kp = basekeyparser.BaseKeyParser(0, supports_count=False,
+                                         supports_chains=False)
+        kp.read_config = mock.Mock()
+
+        kp._modename = 'normal'
+        kp.on_keyconfig_changed('normal2')
+        # Modenames are not equal so read_config() should not be called
+        assert not kp.read_config.called
+
+        kp.on_keyconfig_changed('normal')
+        # Both modenames equal so read_config() should be called
+        assert kp.read_config.called
+
+    def test_warn_on_keychains(self, fake_keyevent_factory, monkeypatch):
+        """Test _warn_on_keychains."""
+        kp = basekeyparser.BaseKeyParser(0, supports_count=False,
+                                         supports_chains=False)
```

Comment:
**Note: This is only an information - you can leave it as-is, as I need to refactor the file anyways because there are old unittest patterns in it.**

This is a good example for how [fixtures](http://pytest.org/latest/fixture.html) are useful - see how you're constructing the same `BaseKeyParser` object in all those tests? You could instead do something like:

```python
@pytest.fixture
def kp(self):
    return basekeyparser.BaseKeyParser(0, supports_count=False, supports_chains=False)

def test_on_keyconfig_changed_mode_none(self, kp):
    assert kp._modename is None
    # ...

def test_on_keyconfig_changed_mode_normal(self, kp):
    kp.read_config = mock.Mock()
    # ...
```